File: vboot/test_keys/gen_keys.py
#!/usr/bin/env python3

import os
import subprocess

def run(*cmd):
    print(' '.join(cmd))
    return subprocess.run(cmd, check=True)


class KeyName:
    def __init__(self, base_name):
        self.base_name = base_name

    def pem(self):
        return self.base_name + '.pem'

    def pub_pem(self):
        return self.base_name + '.pub.pem'

    def vbprivk(self):
        return self.base_name + '.vbprivk'

    def vbpubk(self):
        return self.base_name + '.vbpubk'

    def keyblock(self):
        return self.base_name + '.keyblock'


def gen_key(key_name):
    file_name = key_name.pem()
    # Key generation takes a while, so don't do it if the key
    # already exists.
    if os.path.exists(file_name):
        print(file_name, 'already exists')
    else:
        run('openssl', 'genrsa', '-F4', '-out', file_name, '8192')
    # Also generate the public key in PEM format. This isn't used by
    # vboot utilities, but is used in the Rust tests.
    run('openssl', 'rsa', '-pubout', '-in', file_name,
        '-out', key_name.pub_pem())


def gen_keypair(futility_path, key_name):
    """Generate .vbprivk and .vbpubk files from a .pem file."""
    run(futility_path, '--vb1', 'create', key_name.pem())


def gen_keyblock(futility_path, kernel_key, kernel_data_key):
    # Copied from vboot_reference/firmware/2lib/include/2struct.h
    VB2_KEYBLOCK_FLAG_DEVELOPER_0 = 0x1
    VB2_KEYBLOCK_FLAG_RECOVERY_0 = 0x4
    flags = VB2_KEYBLOCK_FLAG_DEVELOPER_0 | VB2_KEYBLOCK_FLAG_RECOVERY_0

    run(futility_path, '--vb1', 'sign',
        '--signprivate', kernel_key.vbprivk(),
        '--flags', str(flags),
        kernel_data_key.vbpubk(),
        kernel_data_key.keyblock())


def main():
    script_dir = os.path.dirname(os.path.realpath(__file__))

    # Run from within the test_keys subdirectory.
    os.chdir(script_dir)

    futility_path = '../../third_party/vboot_reference/build/futility/futility'

    # Key used to sign the kernel keyblock which contains the public
    # part of the kernel_data_key.
    kernel_key = KeyName('kernel_key')
    gen_key(kernel_key)

    # Key used to sign the kernel data.
    kernel_data_key = KeyName('kernel_data_key')
    gen_key(kernel_data_key)

    gen_keypair(futility_path, kernel_key)
    gen_keypair(futility_path, kernel_data_key)
    gen_keyblock(futility_path, kernel_key, kernel_data_key)


if __name__ == '__main__':
    main()

File: xtask/src/gen_keys.rs
use crate::config::Config;
use crate::sign::KeyPaths;
use anyhow::{Context, Error};
use camino::Utf8Path;
use command_run::Command;
use fehler::throws;

#[throws]
fn gen_key(key_paths: &KeyPaths) {
    let path = key_paths.priv_pem();
    // Key generation takes a while, so don't do it if the key
    // already exists.
    if !path.exists() {
        Command::with_args(
            "openssl",
            &["genrsa", "-F4", "-out", path.as_str(), "8192"],
        )
        .run()?;
    }
    // Also generate the public key in PEM format. This isn't used by
    // vboot utilities, but is used in the Rust tests.
    Command::with_args(
        "openssl",
        &[
            "rsa",
            "-pubout",
            "-in",
            path.as_str(),
            "-out",
            key_paths.pub_pem().as_str(),
        ],
    )
    .run()?;
}

/// Generate .vbprivk and .vbpubk files from a .pem file.
#[throws]
fn gen_keypair(futility_path: &Utf8Path, key_paths: &KeyPaths) {
    Command::with_args(
        futility_path,
        &[
            "--vb1",
            "create",
            key_paths.priv_pem().as_str(),
            key_paths.base_name_path().as_str(),
        ],
    )
    .run()?;
}

#[throws]
fn gen_keyblock(
    futility_path: &Utf8Path,
    kernel_key: &KeyPaths,
    kernel_data_key: &KeyPaths,
) {
    // Copied from vboot_reference/firmware/2lib/include/2struct.h
    const VB2_KEYBLOCK_FLAG_DEVELOPER_0: u32 = 0x1;
    const VB2_KEYBLOCK_FLAG_RECOVERY_0: u32 = 0x4;
    const VB2_KEYBLOCK_FLAG_MINIOS_0: u32 = 0x10;
    let flags = VB2_KEYBLOCK_FLAG_DEVELOPER_0
        | VB2_KEYBLOCK_FLAG_RECOVERY_0
        | VB2_KEYBLOCK_FLAG_MINIOS_0;

    Command::with_args(
        futility_path,
        &[
            "--vb1",
            "sign",
            "--signprivate",
            kernel_key.vbprivk().as_str(),
            "--flags",
            &flags.to_string(),
            kernel_data_key.vbpubk().as_str(),
            kernel_data_key.keyblock().as_str(),
        ],
    )
    .run()?;
}

#[throws]
pub fn generate_test_keys(conf: &Config) {
    let futility_path = &conf.futility_executable_path();
    let kernel_key = &conf.kernel_key_paths();
    let kernel_data_key = &conf.kernel_data_key_paths();

    kernel_key.create_dir()?;
    kernel_data_key.create_dir()?;

    gen_key(kernel_key)?;
    gen_key(kernel_data_key)?;

    gen_keypair(futility_path, kernel_key).context("gen kernel_key failed")?;
    gen_keypair(futility_path, kernel_data_key)
        .context("gen kernel_data_key failed")?;
    gen_keyblock(futility_path, kernel_key, kernel_data_key)
        .context("gen_keyblock failed")?;
}
